src/combinatorics/simple_graph/handshake.lean

```diff
@@ -0,0 +1,262 @@
+begin
+  rw ←card_univ,
+  rw @card_eq_sum_card_fiberwise _ _ _ dart.edge _ G.edge_finset
+    (λ d h, by { rw mem_edge_finset, apply dart.edge_mem }),
+  rw [←mul_comm, sum_const_nat],
+  intros e h,
+  apply G.dart_edge_fiber_card e,
+  rwa ←mem_edge_finset,
+end
+
+/-- The degree-sum formula.  This is also known as the handshaking lemma, which might
+more specifically refer to `simple_graph.card_odd_degree_vertices_is_even`. -/
+theorem sum_degrees_eq_twice_card_edges : ∑ v, G.degree v = 2 * G.edge_finset.card :=
+G.dart_card_eq_sum_degrees.symm.trans G.dart_card_eq_twice_card_edges
+
+end degree_sum
+
+
+section TODO_move
+
+lemma zmod_eq_zero_iff_even (n : ℕ) : (n : zmod 2) = 0 ↔ even n :=
+(char_p.cast_eq_zero_iff (zmod 2) 2 n).trans even_iff_two_dvd.symm
+
+lemma zmod_eq_one_iff_odd (n : ℕ) : (n : zmod 2) = 1 ↔ odd n :=
+begin
+  change (n : zmod 2) = ((1 : ℕ) : zmod 2) ↔ _,
+  rw [zmod.eq_iff_modeq_nat, nat.odd_iff],
+  trivial,
+end
+
+lemma zmod_ne_zero_iff_odd (n : ℕ) : (n : zmod 2) ≠ 0 ↔ odd n :=
+by split; { contrapose, simp [zmod_eq_zero_iff_even], }
+
+end TODO_move
+
+/-- The handshaking lemma.  See also `simple_graph.sum_degrees_eq_twice_card_edges`. -/
+theorem card_odd_degree_vertices_is_even [fintype V] :
+  even (univ.filter (λ v, odd (G.degree v))).card :=
+begin
+  classical,
+  have h := congr_arg ((λ n, ↑n) : ℕ → zmod 2) G.sum_degrees_eq_twice_card_edges,
+  simp only [zmod.cast_self, zero_mul, nat.cast_mul] at h,
+  rw sum_nat_cast at h,
+  rw ←sum_filter_ne_zero at h,
+  rw @sum_congr _ (zmod 2) _ _ (λ v, (G.degree v : zmod 2)) (λ v, (1 : zmod 2)) _ rfl at h,
+  { simp only [filter_congr_decidable, mul_one, nsmul_eq_mul, sum_const, ne.def] at h,
+    rw ←zmod_eq_zero_iff_even,
+    convert h,
+    ext v,
+    rw ←zmod_ne_zero_iff_odd,
+    congr' },
+  { intros v,
+    simp only [true_and, mem_filter, mem_univ, ne.def],
+    rw [zmod_eq_zero_iff_even, zmod_eq_one_iff_odd, nat.odd_iff_not_even, imp_self],
+    trivial }
+end
+
+lemma card_odd_degree_vertices_ne_is_odd [fintype V] [decidable_eq V]
+  (v : V) (h : odd (G.degree v)) :
+  odd (univ.filter (λ w, w ≠ v ∧ odd (G.degree w))).card :=
+begin
+  rcases G.card_odd_degree_vertices_is_even with ⟨k, hg⟩,
+  have hk : 0 < k,
+  { have hh : (filter (λ (v : V), odd (G.degree v)) univ).nonempty,
+    { use v,
+      simp only [true_and, mem_filter, mem_univ],
+      use h, },
+    rw [←card_pos, hg] at hh,
+    clear hg,
+    linarith, },
+  have hc : (λ (w : V), w ≠ v ∧ odd (G.degree w)) = (λ (w : V), odd (G.degree w) ∧ w ≠ v),
+  { ext w,
+    rw and_comm, },
+  simp only [hc, filter_congr_decidable],
+  rw [←filter_filter, filter_ne', card_erase_of_mem],
+  { use k - 1,
+    rw [nat.pred_eq_succ_iff, hg],
+    clear hc hg,
+    rw nat.mul_sub_left_distrib,
+    omega, },
+  { simpa only [true_and, mem_filter, mem_univ] },
+end
+
+lemma exists_ne_odd_degree_if_exists_odd [fintype V]
+  (v : V) (h : odd (G.degree v)) :
```

Comment:
Maybe this is too long...?
```suggestion
lemma exists_ne_odd_degree_of_exists_odd_degree_vertex [fintype V]
  {v : V} (h : odd (G.degree v)) :
```

Reply:
I went with `exists_ne_odd_degree_of_exists_odd_degree`.  Naming theory: write `degree_vertices` if plural, and `degree` if singular?

Reply:
That was my second choice, so LGTM!